File: 2022/02/rust/main.rs
use std::fs::File;
use std::io::{ BufRead, BufReader };
use std::str::FromStr;

const INPUT_FILE:&'static str = "../input/input.txt";

//
// This enumeration represents the result of a game.
//
#[derive(Debug)]
enum GameResult { Win, Draw, Lose }

impl FromStr for GameResult {
    type Err = std::string::ParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let element = match s {
            "X" => GameResult::Lose,
            "Y" => GameResult::Draw,
            "Z" => GameResult::Win,
            _ => panic!("Illegal argument {}", s)
        };
        Ok(element)
    }
}

//
// This enumeration represents the shapes of the game: rock, paper, scissors.
//
#[derive(PartialEq, Clone, Debug)]
enum Shape { Rock, Paper, Scissors }

impl Shape {
    // This method returns the shape to win, lose or draw against this shape.
    pub fn shape_to(&self, result: &GameResult) -> Shape {
        match result {
            GameResult::Win => match self {
                Shape::Rock => Shape::Paper,
                Shape::Paper => Shape::Scissors,
                Shape::Scissors => Shape::Rock
            },
            GameResult::Lose => match self {
                Shape::Rock => Shape::Scissors,
                Shape::Paper => Shape::Rock,
                Shape::Scissors => Shape::Paper
            },
            GameResult::Draw => self.clone()
        }
    }

    // This method returs the points gained for using this shape.
    fn points(&self) -> usize {
        match self {
            Shape::Rock => 1,
            Shape::Paper => 2,
            Shape::Scissors => 3
        }
    }

    // This method matches this shape agains another  and returns a result.
    fn do_match(&self, opponent: &Shape) -> GameResult {
        if self == opponent { return GameResult::Draw; }
        if (self == &Shape::Rock && opponent == &Shape::Scissors) ||
            (self == &Shape::Scissors && opponent == &Shape::Paper) ||
            (self == &Shape::Paper && opponent == &Shape::Rock) { return GameResult::Win; }
        GameResult::Lose
    }

    // This method plays the game against this shape.
    pub fn play(&self, opponent: &Shape) -> usize {
        self.points() + match self.do_match(opponent) {
            GameResult::Win => 6,
            GameResult::Draw => 3,
            GameResult::Lose => 0
        }
    }
}

impl FromStr for Shape {
    type Err = std::string::ParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let element = match s {
            "A" | "X" => Shape::Rock,
            "B" | "Y" => Shape::Paper,
            "C" | "Z" => Shape::Scissors,
            _ => panic!("Illegal argument {}", s)
        };
        Ok(element)
    }
}

fn main() -> Result<(), std::io::Error> {
    // --- Puzzle: Part 1 ---
    let file = File::open(INPUT_FILE)?;
    let total: usize = BufReader::new(file).lines().map(|line| {
        let line = line.unwrap();
        let op: Shape = line[0..1].parse().unwrap();
        let me: Shape = line[2..3].parse().unwrap();
        me.play(&op)
    }).sum();
    println!("Part 1. Total score: {}", total);

    // --- Puzzle: Part 2 ---
    let file = File::open(INPUT_FILE)?;
    let total: usize = BufReader::new(file).lines().map(|line| {
        let line = line.unwrap();
        let op: Shape = line[0..1].parse().unwrap();
        let result: GameResult = line[2..3].parse().unwrap();
        let me: Shape = op.shape_to(&result);
        me.play(&op)
    }).sum();
    println!("Part 2. Total score: {}", total);

    Ok(())
}

File: 2022/02/python/main.py
from enum import Enum

# --- Constants ---
INPUT_FILE = "../input/input.txt"

# --- Enums ----
class GameResult(Enum):
    LOSE = 0
    DRAW = 3
    WIN = 6

    @staticmethod
    def parse(value):
        if value == "X":
            return GameResult.LOSE
        elif value == "Y":
            return GameResult.DRAW
        elif value == "Z":
            return GameResult.WIN
        else:
            return None

class Shape(Enum):
    ROCK = 1
    PAPER = 2
    SCISSORS = 3

    @staticmethod
    def parse(value):
        if value == "A" or value == "X":
            return Shape.ROCK
        elif value == "B" or value == "Y":
            return Shape.PAPER
        elif value == "C" or value == "Z":
            return Shape.SCISSORS
        else:
            return None

# --- Plays a game ---
def play(p1, p2):
    if p1 == Shape.ROCK:
        if p2 == Shape.ROCK:
            return GameResult.DRAW
        elif p2 == Shape.PAPER:
            return GameResult.LOSE
        elif p2 == Shape.SCISSORS:
            return GameResult.WIN
    elif p1 == Shape.PAPER:
        if p2 == Shape.ROCK:
            return GameResult.WIN
        elif p2 == Shape.PAPER:
            return GameResult.DRAW
        elif p2 == Shape.SCISSORS:
            return GameResult.LOSE
    elif p1 == Shape.SCISSORS:
        if p2 == Shape.ROCK:
            return GameResult.LOSE
        elif p2 == Shape.PAPER:
            return GameResult.WIN
        elif p2 == Shape.SCISSORS:
            return GameResult.DRAW
    else:
        None

# --- Selects what to play ---
def choose(p, result):
    if p == Shape.ROCK:
        if result == GameResult.WIN:
            return Shape.PAPER
        elif result == GameResult.DRAW:
            return Shape.ROCK
        elif result == GameResult.LOSE:
            return Shape.SCISSORS
    elif p == Shape.PAPER:
        if result == GameResult.WIN:
            return Shape.SCISSORS
        elif result == GameResult.DRAW:
            return Shape.PAPER
        elif result == GameResult.LOSE:
            return Shape.ROCK
    elif p == Shape.SCISSORS:
        if result == GameResult.WIN:
            return Shape.ROCK
        elif result == GameResult.DRAW:
            return Shape.SCISSORS
        elif result == GameResult.LOSE:
            return Shape.PAPER

# --- Read the input file ---
total_1 = 0
total_2 = 0
file = open(INPUT_FILE)
while True:
    line = file.readline()
    if not line:
        break

    line = line.strip()

    # --- Puzzle 1 ---
    elf = Shape.parse(line[0])
    me = Shape.parse(line[2])
    result = play(me, elf)
    total_1 += result.value + me.value

    # --- Puzzle 2 ---
    result = GameResult.parse(line[2])
    me = choose(elf, result)
    total_2 += result.value + me.value


print(f"1. Total score: {total_1:,d}")
print(f"2. Total score: {total_2:,d}")
